ql/pricingengines/credit/isdacdsengine.cpp

```diff
@@ -0,0 +1,395 @@
+        // but the ISDA engine is not explicitly specified to handle them,
+        // so we just forbid them too
+
+        Actual365Fixed dc;
+        Actual360 dc1;
+        Actual360 dc2(true);
+
+        Date evalDate = Settings::instance().evaluationDate();
+
+        // check if given curves are ISDA compatible
+        // (the interpolation is checked below)
+
+        QL_REQUIRE(!discountCurve_.empty(), "no discount term structure set");
+        QL_REQUIRE(!probability_.empty(), "no probability term structure set");
+        QL_REQUIRE(discountCurve_->dayCounter() == dc,
+                   "yield term structure day counter ("
+                       << discountCurve_->dayCounter()
+                       << ") should be Act/365(Fixed)");
+        QL_REQUIRE(probability_->dayCounter() == dc,
+                   "probability term structure day counter ("
+                       << probability_->dayCounter() << ") should be "
+                       << "Act/365(Fixed)");
+        QL_REQUIRE(discountCurve_->referenceDate() == evalDate,
+                   "yield term structure reference date ("
+                       << discountCurve_->referenceDate()
+                       << " should be evaluation date (" << evalDate << ")");
+        QL_REQUIRE(probability_->referenceDate() == evalDate,
+                   "probability term structure reference date ("
+                       << probability_->referenceDate()
+                       << " should be evaluation date (" << evalDate << ")");
+        QL_REQUIRE(arguments_.settlesAccrual,
+                   "ISDA engine not compatible with non accrual paying CDS");
+        QL_REQUIRE(arguments_.paysAtDefaultTime,
+                   "ISDA engine not compatible with end period payment");
+        QL_REQUIRE(boost::dynamic_pointer_cast<FaceValueClaim>(
+                       arguments_.claim) != NULL,
+                   "ISDA engine not compatible with non face value claim");
+
+        Date maturity = arguments_.maturity;
+        Date effectiveProtectionStart =
+            std::max<Date>(arguments_.protectionStart, evalDate + 1);
+
+        // collect nodes from both curves and sort them
+        std::vector<Date> yDates, cDates;
+
+        if(boost::shared_ptr<PiecewiseYieldCurve<Discount, LogLinear> > castY1 =
+            boost::dynamic_pointer_cast<
+                PiecewiseYieldCurve<Discount, LogLinear> >(*discountCurve_)) {
+            yDates = castY1->dates();
+        } else if(boost::shared_ptr<PiecewiseYieldCurve<ForwardRate, BackwardFlat> >
+        castY2 = boost::dynamic_pointer_cast<
+            PiecewiseYieldCurve<ForwardRate, BackwardFlat> >(*discountCurve_)) {
+            yDates = castY2->dates();
+        } else if(boost::shared_ptr<FlatForward> castY3 =
+            boost::dynamic_pointer_cast<FlatForward>(*discountCurve_)) {
+        } else {
+            QL_FAIL("Yield curve must be flat forward interpolated");
```

Comment:
I would check for the base class `InterpolatedDiscountCurve<LogLinear>` instead of `PiecewiseYieldCurve<Discount, LogLinear>`.  This allows the user to use the former and pass the rates instead of bootstrapping them, or to use `PiecewiseYieldCurve` with a different bootstrap algorithm, or even to inherit their own interpolated term structure.  Also, you can probably allow `ForwardFlat` as well as `BackwardFlat`.